src/agent/include/register.hpp

```diff
@@ -11,5 +19,21 @@ namespace registration
         std::string password;
     };
 
-    bool RegisterAgent(const UserCredentials& userCredentials);
+    bool RegisterAgent(
+        const UserCredentials& userCredentials,
+        std::function<std::optional<std::string>(
+            const std::string&, const std::string&, const std::string&, const std::string&)> AuthenticateFunction,
```

Comment:
Though parameter names are not needed here, given that amount of arguments, we should better include them so readers of this function signature know what it's about. It's also weird to have RegistrationFunction with each argument named, but AuthenticateFunction with none.

Furthermore, since the `RegisterAgent` declaration is a bit too verbose we could try to use type alises or decltype to make it more legible.

```c++

    using RegistrationFunctionType = http::status(const std::string& host,
                                                  const std::string& port,
                                                  const std::string& token,
                                                  const std::string& uuid,
                                                  const std::string& key,
                                                  const std::optional<std::string>& name);

    RegistrationFunctionType SendRegistrationRequest;

    bool RegisterAgent(
        const UserCredentials& userCredentials,
        std::function<std::optional<std::string>(
            const std::string&, const std::string&, const std::string&, const std::string&)> AuthenticateFunction,
        std::function<RegistrationFunctionType> RegistrationFunction);
```

The same could be done for AuthenticateFunction, but the function type/signature, should be declared in http_client.

Reply:
Also, nitpicking a bit here, but could we make the names "symmetrical"? Ie: AuthenticateFunction and RegisterFunction, or AuthenticationFunction and RegistrationFunction.

Reply:
Done!